### PR TITLE
Refactoring prometheus module to aggregate metrics based on metric family

### PR DESCRIPTION
**Pull request by vjsamuel**
The current prometheus collector implementation separates each metric into a separate event. This is not how prometheus is meant to be understood. Prometheus has the concept of MetricFamily.

Example of a metric family would be: 
```
# TYPE apiserver_request_latencies histogram
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="125000"} 11542
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="250000"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="500000"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="1e+06"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="2e+06"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="4e+06"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="8e+06"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="GET",le="+Inf"} 11543
apiserver_request_latencies_sum{resource="accounts",verb="GET"} 1.7285094e+07
apiserver_request_latencies_count{resource="accounts",verb="GET"} 11543
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="125000"} 566
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="250000"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="500000"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="1e+06"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="2e+06"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="4e+06"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="8e+06"} 567
apiserver_request_latencies_bucket{resource="accounts",verb="LIST",le="+Inf"} 567
apiserver_request_latencies_sum{resource="accounts",verb="LIST"} 4.695524e+06
apiserver_request_latencies_count{resource="accounts",verb="LIST"} 567
 ```
This metric family can be broken down as:
`metric_name: apiserver_request_latencies`
`metric_type: histogram`

It has two different label combinations:
`{resource="accounts",verb="LIST"}` and `{resource="accounts",verb="GET"}`

Hence two events ought to be created. one of which would look similar to:
```
      "apiserver_request_latencies": {
        "buckets": {
          "+Inf": 3696,
          "1000000": 0,
          "125000": 0,
          "2000000": 0,
          "250000": 0,
          "4000000": 0,
          "500000": 0,
          "8000000": 0
        },
        "count": 3696,
        "sum": 1668259775362.000000
      },
      "labels": {
        "resource": "roles",
        "verb": "WATCHLIST"
      }
    }
  }
```

This enables histograms and summaries to be looked adjacent to each other.

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -78,6 +78,8 @@ https://github.com/elastic/beats/compare/v5.1.1...master[Check the HEAD diff]
 - Make system process metricset honor the cpu_ticks config option. {issue}3590[3590]
 - Support common.Time in mapstriface.toTime() {pull}3812[3812]
 - Fixing panic on prometheus collector when label has , {pull}3947[3947]
+- Fixing prometheus collector to aggregate metrics based on metric family. {pull}4075[4075]
```
Comment (ruflin) [inline CHANGELOG.asciidoc]: I would not put this under bugfixes but breaking changes as it changes the data structure.

File: metricbeat/module/prometheus/collector/collector.go [hunk 2]
```diff
@@ -6,6 +6,9 @@ import (
 	"github.com/elastic/beats/metricbeat/helper"
 	"github.com/elastic/beats/metricbeat/mb"
 	"github.com/elastic/beats/metricbeat/mb/parse"
+
+	"fmt"
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/collector.go]: could you move the fmt on the top to have the standard imports together?

File: metricbeat/module/prometheus/collector/collector.go [hunk 3]
```diff
@@ -53,36 +56,36 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+				if len(promEvent.labels) > 0 {
+					eventList[promEvent.labelHash]["labels"] = promEvent.labels
+				}
 
-			// Add labels
-			if len(promEvent.labels) > 0 {
-				eventList[promEvent.labelHash]["label"] = promEvent.labels
 			}
 
+			eventList[promEvent.labelHash][promEvent.key] = promEvent.value
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/collector.go]: I assume the promEvent.key are quite constant over time so we don't have a field explosion here.

File: metricbeat/module/prometheus/collector/collector_test.go [hunk 4]
```diff
@@ -5,71 +5,138 @@ package collector
...
+				},
+			},
 			Event: PromEvent{
-				key:       "http_request_duration_microseconds",
-				value:     int64(17),
-				labelHash: `handler="query",quantile="0.99"`,
-				labels: common.MapStr{
-					"handler":  "query",
-					"quantile": 0.99,
+				key: "http_request_duration_microseconds",
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/collector_test.go]: In a future step we could also extract the "units" part like `microseconds` as I assume this is also part of the convention.

File: metricbeat/module/prometheus/collector/data.go [hunk 5]
```diff
@@ -1,83 +1,94 @@
 package collector
 
 import (
-	"strconv"
-	"strings"
-
 	"github.com/elastic/beats/libbeat/common"
+	dto "github.com/prometheus/client_model/go"
+	"math"
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: can you move these 2 to the top?

Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: Our general logic is for imports:

```
standard imports

beats imports

external imports
```

File: metricbeat/module/prometheus/collector/data.go [hunk 6]
```diff
@@ -1,83 +1,94 @@
...
-		keyValue[1] = strings.Trim(keyValue[1], "\"")
+func GetPromEventsFromMetricFamily(mf *dto.MetricFamily) []PromEvent {
+	var events []PromEvent
+
+	name := *mf.Name
+	metrics := mf.Metric
+	for _, metric := range metrics {
+		event := PromEvent{
+			key:       name,
+			labelHash: "#",
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: How does that exactly work?

File: metricbeat/module/prometheus/collector/data.go [hunk 7]
```diff
@@ -1,83 +1,94 @@
...
+			value["count"] = summary.GetSampleCount()
+
+			quantiles := summary.GetQuantile()
+
+			percentileMap := common.MapStr{}
+			for _, quantile := range quantiles {
+				key := strconv.FormatFloat((100 * quantile.GetQuantile()), 'f', -1, 64)
+
+				if math.IsNaN(quantile.GetValue()) == false {
+					percentileMap["p"+key] = quantile.GetValue()
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: As it is already under `percentile` I don't think we ned to add a `p` in front of the key.

File: metricbeat/module/prometheus/collector/data.go [hunk 8]
```diff
@@ -1,83 +1,94 @@
...
+				}
+
+			}
+
+			if len(percentileMap) != 0 {
+				value["percentiles"] = percentileMap
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: Lets make it `percentile` as then it reads `precentile.99: 0.2`

File: metricbeat/module/prometheus/collector/data.go [hunk 9]
```diff
@@ -1,83 +1,94 @@
...
+			value["sum"] = histogram.GetSampleSum()
+			value["count"] = histogram.GetSampleCount()
+			buckets := histogram.GetBucket()
+			bucketMap := common.MapStr{}
+			for _, bucket := range buckets {
+				key := strconv.FormatFloat(bucket.GetUpperBound(), 'f', -1, 64)
+				bucketMap[key] = bucket.GetCumulativeCount()
+			}
+
+			value["buckets"] = bucketMap
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/data.go]: Same here, lets make it `bucket`

File: metricbeat/module/prometheus/utils.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
+package prometheus
+
+import (
+	"fmt"
+	dto "github.com/prometheus/client_model/go"
```
Comment (ruflin) [inline metricbeat/module/prometheus/utils.go]: see import rules above

File: metricbeat/module/prometheus/collector/_meta/data.json [hunk 11]
```diff
@@ -13,11 +13,13 @@
     },
     "prometheus": {
         "collector": {
-            "label": {
+            "labels": {
```
Comment (ruflin) [inline metricbeat/module/prometheus/collector/_meta/data.json]: I would prefer to have here `label` even though I'm aware this is not consitent with most of the other label fields with have. They should also be singular.

File: metricbeat/module/prometheus/collector/collector.go (hunk 3 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/collector.go]: +1

File: metricbeat/module/prometheus/collector/data.go (hunk 6 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/data.go]: this makes sure that all metrics that dont have any tag values gets grouped into a single document. its a carry over from the previous implementation that was there before this change.

File: metricbeat/module/prometheus/collector/_meta/data.json (hunk 11 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/_meta/data.json]: done

File: metricbeat/module/prometheus/utils.go (hunk 10 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/utils.go]: done.

File: metricbeat/module/prometheus/collector/data.go (hunk 9 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/data.go]: done

File: metricbeat/module/prometheus/collector/data.go (hunk 8 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/data.go]: done

File: metricbeat/module/prometheus/collector/data.go (hunk 7 above)
Comment (vjsamuel) [inline metricbeat/module/prometheus/collector/data.go]: done

File: CHANGELOG.asciidoc [hunk 12]
```diff
@@ -79,6 +80,7 @@ https://github.com/elastic/beats/compare/v5.1.1...master[Check the HEAD diff]
 - Support common.Time in mapstriface.toTime() {pull}3812[3812]
 - Fixing panic on prometheus collector when label has , {pull}3947[3947]
 - Fix MongoDB dbstats fields mapping. {pull}4025[4025]
+- Fixing prometheus collector to aggregate metrics based on metric family. {pull}4075[4075]
```
Comment (ruflin) [inline CHANGELOG.asciidoc]: This one should now be removed.

**Comment by tsg** [comment]
@ruflin @vjsamuel The new document organization makes a lot of sense to me, and it's great we can group more metrics together. 👍  👍  👍

**Pull request merged**